### PR TITLE
PUB-727 - Ingest of Publications

### PR DESCRIPTION
**Pull request by ChrisS1512**
### JIRA link (if applicable) ###

https://tools.hmcts.net/jira/browse/PUB-727

### Change description ###

This PR captures the ingest of publications into P&I.

It does three things:

1) Stores the artefact metadata in Postgres
2) Stores the artefact payload in Blob storage
3) Provides a reference to the blob in the Postgres database
4) Extracts search values into the search field, in JSON format.

See README.md for details of the environment variables that need setting up.

Documentation will be written around the mandatory fields, types etc and how the search works.

The payload searching is fully configurable in the application.yaml. Currently it only parses JSON payloads, however we can extend with further parsers in the future.

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

### CODE REVIEW

File: src/main/java/uk/gov/hmcts/reform/pip/data/management/database/AzureBlobService.java [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+    /**
+     * Creates the payload in the Azure blob service.
+     * @param sourceArtefactId The source ID to call the blob.
+     * @param provenance The provenance to call the blob.
+     * @param payload The payload to create
+     * @return The URL to where the payload has been created
+     */
+    public String createPayload(String sourceArtefactId, String provenance, String payload) {
+        String blobName = sourceArtefactId + '-' + provenance;
+        BlobClient blobClient = blobContainerClient.getBlobClient(sourceArtefactId + '-' + provenance);
```
Comment (kurt-uk) [inline src/main/java/uk/gov/hmcts/reform/pip/data/management/database/AzureBlobService.java:32]:
```suggestion
        BlobClient blobClient = blobContainerClient.getBlobClient(blobName);
```
?

**Review by kurt-uk — CHANGES_REQUESTED** [review]

File: src/main/java/uk/gov/hmcts/reform/pip/data/management/database/AzureBlobService.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (ChrisS1512) [reply src/main/java/uk/gov/hmcts/reform/pip/data/management/database/AzureBlobService.java]: Nice spot! Updated

File: src/main/java/uk/gov/hmcts/reform/pip/data/management/errorhandling/GlobalExceptionHandler.java [hunk 2]
```diff
@@ -38,4 +41,35 @@
...
+    }
+
+    @ExceptionHandler(EmptyRequestHeaderException.class)
+    public ResponseEntity<ExceptionResponse> handle(EmptyRequestHeaderException ex) {
+
+        ExceptionResponse exceptionResponse = new ExceptionResponse();
+        exceptionResponse.setMessage(ex.getMessage());
+        exceptionResponse.setTimestamp(LocalDateTime.now());
+
+        return ResponseEntity.status(HttpStatus.BAD_REQUEST).body(exceptionResponse);
```
[the remark this answers is not in the document]
Comment (ChrisS1512) [reply src/main/java/uk/gov/hmcts/reform/pip/data/management/errorhandling/GlobalExceptionHandler.java:72]: Just had a look at this. I tried it out however the spring exception has a complex constructor which takes a spring MethodParameter which I couldn't see an easy way to create/construct so we can pass it in, therefore I have left this as it is.

File: src/main/java/uk/gov/hmcts/reform/pip/data/management/models/publication/Artefact.java [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+import javax.persistence.Column;
+import javax.persistence.Entity;
+import javax.persistence.EnumType;
+import javax.persistence.Enumerated;
+import javax.persistence.GeneratedValue;
+import javax.persistence.GenerationType;
+import javax.persistence.Id;
+
+/**
+ * Class that presents the Inbound artifact that is being published.
```
[the remark this answers is not in the document]
Comment (ChrisS1512) [reply src/main/java/uk/gov/hmcts/reform/pip/data/management/models/publication/Artefact.java]: Updated :)

File: src/test/java/uk/gov/hmcts/reform/pip/data/management/controllers/PublicationControllerTest.java [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+                publicationController.uploadPublication(
+                    PROVENANCE, EMPTY_FIELD, ARTEFACT_TYPE,
+                    SENSITIVITY, LANGUAGE, DISPLAY_FROM, DISPLAY_TO, PAYLOAD
+                );
+            });
+
+        assertEquals("x-source-artefact-id is mandatory however an empty value is provided",
+                     emptyRequestHeaderException.getMessage(),
+                     "The expected exception does not contain the correct message");
+    }
```
[the remark this answers is not in the document]
Comment (ChrisS1512) [reply src/test/java/uk/gov/hmcts/reform/pip/data/management/controllers/PublicationControllerTest.java:129]: Added a new test to make sure it stops on the first.

In terms of the test empty type, that's captured by the spring validation therefore I've just tested this one in PublicationTest in the integration tests.

File: src/main/java/uk/gov/hmcts/reform/pip/data/management/errorhandling/GlobalExceptionHandler.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (kurt-uk) [reply src/main/java/uk/gov/hmcts/reform/pip/data/management/errorhandling/GlobalExceptionHandler.java:72]: ok 😊

**Review by kurt-uk — APPROVED** [review]

**Review by commitBlob — APPROVED** [review]

File: src/main/java/uk/gov/hmcts/reform/pip/data/management/controllers/PublicationController.java [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+     * @param type List / Outcome / Judgement / Status Updates.
+     * @param sensitivity Level of sensitivity.
+     * @param language Language of publication.
+     * @param displayFrom Date / Time from which the publication will be displayed.
+     * @param displayTo Date / Time until which the publication will be displayed.
+     * @param payload JSON Blob with key/value pairs of data to be published.
+     * @return The created artefact.
+     */
+    @ApiResponses({
+        @ApiResponse(code = 201, message = "Publication has been created"),
```
Comment (kurt-uk) [inline src/main/java/uk/gov/hmcts/reform/pip/data/management/controllers/PublicationController.java:58]: it looks like this message isnt returned, you return the artifact object with no message

Comment (ChrisS1512) [reply src/main/java/uk/gov/hmcts/reform/pip/data/management/controllers/PublicationController.java]: Good spot! Updated

**Review by kurt-uk — APPROVED** [review]

**Review by commitBlob — DISMISSED** [review]

**Review by furnivall — APPROVED** [review]
Looks good, nice one :)

**Review by commitBlob — APPROVED** [review]

**Pull request merged**